### PR TITLE
Uri encode settings

### PR DESCRIPTION
**Pull request by rcoh**
Some AWS services require double-encoding URIs.

### CODE REVIEW

File: aws-sigv4/src/types.rs [hunk 1]
```diff
@@ -52,6 +60,35 @@ impl CanonicalRequest {
...
+    .add(b',')
+    .add(b';')
+    .add(b'=')
+    .add(b'%');
+
+//:/?#[]@!$&'()*+,;=
+
+fn double_encode(path: &str) -> String {
+    utf8_percent_encode(path, PATH).to_string()
+}
```
Comment (davidbarsky) [inline aws-sigv4/src/types.rs]: no need to have a helper/standalone function for this—i'd just inline this

File: aws-sigv4/src/types.rs [hunk 2]
```diff
@@ -52,6 +60,35 @@ impl CanonicalRequest {
     }
 }
 
+use percent_encoding::{utf8_percent_encode, AsciiSet, CONTROLS};
+
+const PATH: &AsciiSet = &CONTROLS
```
Comment (davidbarsky) [inline aws-sigv4/src/types.rs]: what's... this doing?

File: aws-sigv4/src/types.rs [hunk 3]
```diff
@@ -52,6 +60,35 @@ impl CanonicalRequest {
     }
 }
 
+use percent_encoding::{utf8_percent_encode, AsciiSet, CONTROLS};
```
Comment (davidbarsky) [inline aws-sigv4/src/types.rs]: move this up to the top of the file?

File: aws-sigv4/src/lib.rs [hunk 4]
```diff
@@ -75,6 +76,24 @@ pub struct Config<'a> {
...
     pub date: SystemTime,
+
+    pub settings: SigningSettings,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct SigningSettings {
+    /// We assume the uri will be encoded once in preparation for transmission.  Certain services
+    /// do not decode before checking signature, requiring us to actually double-encode the uri
+    /// in the canonical request in order to pass a signature check.
```
Comment (davidbarsky) [inline aws-sigv4/src/lib.rs]:
```suggestion
    /// We assume the URI will be encoded _once_ prior to transmission. Some services
    /// do not decode the path prior to checking the signature, requiring clients to actually _double-encode_
    /// the URI in creating the canonical request in order to pass a signature check.
```

Nit: what services?

File: aws-sigv4/src/lib.rs [hunk 5]
```diff
@@ -75,6 +76,24 @@ pub struct Config<'a> {
...
+    double_uri_encode: bool,
```
Comment (davidbarsky) [inline aws-sigv4/src/lib.rs]: Maybe make this an a non-exhaustive enum describing what the behavior should be, instead of a boolean?

**Review by davidbarsky — CHANGES_REQUESTED** [review]

File: aws-sigv4/src/lib.rs (hunk 4 above)
Comment (rcoh) [reply aws-sigv4/src/lib.rs]: there is no conclusive list 💀

File: aws-sigv4/src/types.rs (hunk 2 above)
Comment (rcoh) [reply aws-sigv4/src/types.rs]: it's how you configure the percent encode crate, sadly.

Comment (davidbarsky) [reply aws-sigv4/src/types.rs]: mind adding that as a comment above this?

**Review by davidbarsky — APPROVED** [review]

**Pull request merged**